commands/ls.go

```diff
@@ -3,142 +3,150 @@ package commands
+	"github.com/docker/buildx/util/formatter"
+	"github.com/docker/buildx/util/progress"
 	"github.com/docker/cli/cli"
 	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/opts"
 	"github.com/moby/buildkit/util/appcontext"
+	"github.com/morikuni/aec"
+	"github.com/olekukonko/tablewriter"
 	"github.com/spf13/cobra"
-	"golang.org/x/sync/errgroup"
 )
 
 type lsOptions struct {
+	filter opts.FilterOpt
+	format string
+	quiet  bool
 }
 
+var lsAllowedFilters = map[string]bool{
+	"name": true,
+}
+
+const lsRawTpl = `name: {{ .Name }}
+current: {{ .Current }}
+driver: {{ .Driver }}
+err: {{ .Err }}
+nodes:
+{{ range .Nodes }}
+  name: {{ .Name }}
+  endpoint: {{ .Endpoint }}
+  platforms: {{ join .Platforms "," }}
+  flags: {{ join .Flags " " }}
+  status: {{ .Status }}
+  err: {{ .Err }}
+{{ end }}
+`
+
 func runLs(dockerCli command.Cli, in lsOptions) error {
 	ctx := appcontext.Context()
 
-	txn, release, err := storeutil.GetStore(dockerCli)
+	filters := in.filter.Value()
+	err := filters.Validate(lsAllowedFilters)
 	if err != nil {
 		return err
 	}
-	defer release()
 
-	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
-	defer cancel()
-
-	ll, err := txn.List()
+	txn, release, err := storeutil.GetStore(dockerCli)
 	if err != nil {
 		return err
 	}
+	defer release()
 
-	builders := make([]*nginfo, len(ll))
-	for i, ng := range ll {
-		builders[i] = &nginfo{ng: ng}
-	}
-
-	list, err := dockerCli.ContextStore().List()
+	bc := builderutil.New(dockerCli, txn)
+	builders, err := bc.LoadBuilders()
 	if err != nil {
 		return err
 	}
-	ctxbuilders := make([]*nginfo, len(list))
-	for i, l := range list {
-		ctxbuilders[i] = &nginfo{ng: &store.NodeGroup{
-			Name: l.Name,
-			Nodes: []store.Node{{
-				Name:     l.Name,
-				Endpoint: l.Name,
-			}},
-		}}
-	}
-
-	builders = append(builders, ctxbuilders...)
-
-	eg, _ := errgroup.WithContext(ctx)
 
-	for _, b := range builders {
-		func(b *nginfo) {
-			eg.Go(func() error {
-				err = loadNodeGroupData(ctx, dockerCli, b)
-				if b.err == nil && err != nil {
-					b.err = err
-				}
-				return nil
-			})
-		}(b)
+	if filters.Len() > 0 {
+		var filtered []*builderutil.Builder
+		for _, b := range builders {
+			if filters.Contains("name") && !filters.Match("name", b.NodeGroup.Name) {
+				continue
+			}
+			filtered = append(filtered, b)
+		}
+		builders = filtered
 	}
 
-	if err := eg.Wait(); err != nil {
-		return err
+	if in.quiet {
+		for _, b := range builders {
+			fmt.Println(b.NodeGroup.Name)
+		}
+		return nil
 	}
 
-	currentName := "default"
-	current, err := storeutil.GetCurrentInstance(txn, dockerCli)
-	if err != nil {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	defer cancel()
+
+	progressMode := progress.PrinterModeAuto
+	if in.format != formatter.FormatTable {
+		progressMode = progress.PrinterModeQuiet
+	}
+	pw := progress.NewPrinter(context.Background(), os.Stderr, progressMode)
+	if err = bc.LoadBuildersData(timeoutCtx, builders, pw); err != nil {
```

Comment:
I don't like this progress on `ls`. It is confusing because the user does not want to load or build anything. Yes, in some cases it may take time to list builders but that is the same with any other command.

If we want to improve this then a simple line like "Checking builders online status...` after a couple of seconds timeout would do. Maybe cancelable with `^D` (shows status as "timeout"). But it has nothing to do with filters.